src/main/java/ch/vorburger/exec/AtomicExecuteResultHandler.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * #%L
+ * ch.vorburger.exec
+ * %%
+ * Copyright (C) 2012 - 2023 Michael Vorburger
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package ch.vorburger.exec;
+
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.commons.exec.ExecuteException;
+import org.apache.commons.exec.ExecuteResultHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An {@link ExecuteResultHandler} which uses a single
+ * {@link java.util.concurrent.atomic.AtomicReference}
+ * instead of three separate volatile fields like the
+ * original {@link org.apache.commons.exec.DefaultExecuteResultHandler} did.
+ *
+ * @see <a href="https://github.com/vorburger/ch.vorburger.exec/issues/108">Issue #108</a>
+ */
+public class AtomicExecuteResultHandler implements ExecuteResultHandler {
+    private static final Logger LOG = LoggerFactory.getLogger(AtomicExecuteResultHandler.class);
+
+    /** The interval polling the result */
+    private static final int SLEEP_TIME_MS = 50;
+
+    private static class Holder {
+        /** The exit value of the finished process */
+        private final Integer exitValue;
+
+        /** Any offending exception */
+        private final ExecuteException exception;
+
+        Holder(int exitValue) {
+            this.exitValue = exitValue;
+            this.exception = null;
+        }
+
+        Holder(ExecuteException e) {
+            this.exception = e;
+            this.exitValue = null;
+        }
+
+        @Override
+        public String toString() {
+            return "exitValue=" + exitValue + ", exception=" + exception;
+        }
+    }
+
+    private final AtomicReference<Holder> holder = new AtomicReference<>();
+
+    @Override
+    public void onProcessComplete(int exitValue) {
+        Holder witness = holder.compareAndExchange(null, new Holder(exitValue));
```

Comment:
This is slightly inconvenient for us–we're still on JDK8 (likely for the next year) so using AtomicReference#compareAndExchange (introduced in JDK9) means that we can't use ch.vorburger.exec anymore w/o forking it.  We've already forked mariadb4j for JDK8 support so it's not hugely inconvenient, but I figured it's worth mentioning.

As an aside, do you prefer to use AtomicReference over VarHandle because it has a simpler API?

Reply:
@mosesn I'm totally open to limiting this project to Java 8 only. Let's pursue this in #117.